### PR TITLE
ENG-4522 Fixed cache issues

### PR DESCRIPTION
**Pull request by zonia3000**

### CODE REVIEW

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractGenericCacheWrapper.java [hunk 1]
```diff
@@ -119,7 +117,7 @@ protected <O> void manage(String key, O object, Action operation) {
             return;
         }
         Cache cache = this.getCache();
-        List<String> codes = (List<String>) this.get(cache, this.getCodesCacheKey(), List.class);
+        List<String> codes = new ArrayList<>(this.get(cache, this.getCodesCacheKey(), List.class));
```
Comment (entgigi) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractGenericCacheWrapper.java]: I would like a separate private method to make explicit/more readable what are you doing (`getCodesCopyFromCacheImmutable`) , just a suggestion

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractGenericCacheWrapper.java [hunk 2]
```diff
@@ -63,17 +64,14 @@ protected void insertAndCleanCache(Cache cache, Map<String, O> objects) {
...
-            }
-            codes.add(key);
+        for (Map.Entry<String, O> entry: objects.entrySet()) {
+            cache.put(cacheKeyPrefix + entry.getKey(), entry.getValue());
+            codes.add(entry.getKey());
         }
         cache.put(codesCacheKey, codes);
-        this.releaseObjects(cache, oldCodes, cacheKeyPrefix);
+        List<String> keysToRelease = oldCodes == null ? null :
+                oldCodes.stream().filter(c -> !codes.contains(c)).collect(Collectors.toList());
```
Comment (entgigi) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractGenericCacheWrapper.java]: codes can they be more than 100 ? if yes, we should generate a map of code to reduce the computational complexity at the previous `for` step

File: engine/src/main/java/com/agiletec/aps/system/services/category/cache/CategoryManagerCacheWrapper.java [hunk 3]
```diff
@@ -166,8 +166,9 @@ public void deleteCategory(String code) {
         }
         List<String> codes = (List<String>) this.get(cache, CATEGORY_CODES_CACHE_NAME, List.class);
```
Comment (entgigi) [inline engine/src/main/java/com/agiletec/aps/system/services/category/cache/CategoryManagerCacheWrapper.java]: I suggest to use a separate method as in the previous comment

File: engine/src/main/java/org/entando/entando/aps/system/services/cache/CacheInfoManager.java [hunk 4]
```diff
@@ -217,7 +220,9 @@ public boolean isExpired(String targetCache, String key) {
             return false;
         }
         if (expirationTime.before(new Date())) {
-            expirationTimes.remove(key);
+            Map<String, Date> newExpirationTimes = new HashMap<>(expirationTimes);
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/cache/CacheInfoManager.java:224]: you should put a log here

File: redis-plugin/src/main/java/org/entando/entando/plugins/jpredis/aps/system/redis/LettuceCache.java [hunk 5]
```diff
@@ -57,15 +59,28 @@ public ValueWrapper get(Object key) {
...
     }
     
     @Override
     public synchronized <T> T get(Object key, Callable<T> valueLoader) {
         logger.warn("Calling remote cache for key {}", key);
         return super.get(key, valueLoader);
     }
-    
+
+    private Object getImmutable(Object value) {
```
Comment (entgigi) [inline redis-plugin/src/main/java/org/entando/entando/plugins/jpredis/aps/system/redis/LettuceCache.java:74]: maybe I'm not understanding but this is not immutable this is a copy, we should use ImmutableList/Map from google collections, so they trigger an error ... or maybe i'm saying something stupid?

**Review by entgigi — CHANGES_REQUESTED** [review]

File: redis-plugin/src/main/java/org/entando/entando/plugins/jpredis/aps/system/redis/LettuceCache.java (hunk 5 above)
Comment (zonia3000) [reply redis-plugin/src/main/java/org/entando/entando/plugins/jpredis/aps/system/redis/LettuceCache.java:74]: Method [copyOf()](https://docs.oracle.com/javase/10/docs/api/java/util/List.html#copyOf(java.util.Collection)) of `java.util.Collection` returns an immutable copy, as Guava `ImmutableList`.

> `copyOf​(Collection<? extends E> coll)` | Returns an unmodifiable List containing the elements of  the given Collection, in its iteration order.

File: engine/src/main/java/org/entando/entando/aps/system/services/cache/CacheInfoManager.java (hunk 4 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/cache/CacheInfoManager.java:224]: done

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractGenericCacheWrapper.java (hunk 2 above)
Comment (zonia3000) [reply engine/src/main/java/com/agiletec/aps/system/common/AbstractGenericCacheWrapper.java]: Used directly the map passed as parameter.

**Review by entgigi — APPROVED** [review]
LGTM

**Pull request merged**